---
title: "TINMAN 16S Analysis with SILVA"
author: "Rachel Rodgers"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=8,
                      fig.height=6,
                      warning = FALSE,
                      message = FALSE)

knitr::opts_knit$set(root.dir = "..")
```

```{r load-libraries, message=FALSE}
library("here")
library("shiny")
library("knitr")
library("janitor")
library("phyloseq")
library("vegan")
library("gridExtra")
library("microbiome")
library("data.table")
library("ggrepel")
library("ggpubr")
library("scales")
library("DT")
library("plotly")
library("FSA")
library("DESeq2")
library("tidyverse")

figPath <- "../analysis/figures/16S/"
resultsPath <- "../analysis/results/16S"

source(here("shared_R_scripts/Helper_Functions.R"))
source(here("shared_R_scripts/BoxPlotHelpers.R"))
source(here("shared_R_scripts/BiomarkerHelpers.R"))

options(shiny.sanitize.errors = FALSE)
```

## Overview {.tabset}

In this analysis, we are interested in investigating changes in the gut 
bacterial microbiome in patients who developed neutropenia after antibiotics
(10 patients), and comparing to matched controls (29 patients). Each patient has
a "pre" and "post" time point. This data set will be analyzed in several
different groupings:

  1. Neutropenic vs Controls in the full cohort (time points combined)
  2. Neutropenic vs Controls in the "pre" and "post" time points only
  3. "Pre" vs "Post" timepoints in the Control or Neutropenic group only
  
Taxonomic assignment to the 16S data was made using DADA2 with the SILVA NR 99 
v138.1 training set.
  
```{r physeqRaw, results='hide'}
physeqRaw <- readRDS(here("../data/RDataObjects/16S/physeqObjects/ps0.silva138_single.RDS"))
physeqRaw # 1823 x 76
```

```{r sampleData, results='hide'}
sampleData <- readRDS(here("../data/RDataObjects/16S/metadataFinal.RDS"))

# do we have the same number of samples in sample data and ps obj?
metadataSamples <- sampleData$sample_name
length(metadataSamples) # 76 samples

psSamples <- sample_names(physeqRaw)
length(psSamples) # 76 samples

# which samples in one set are missing from the other:
  # in the physeq object but not in the metadata:
missingFromMetadata <- psSamples[!psSamples %in% metadataSamples]
missingFromMetadata
  # in the metadata but not in the physeq object:
missingFromPS <- metadataSamples[!metadataSamples %in% psSamples]
missingFromPS

# For TINMANVCH* sample names, add underscores.
#   Then, modify the remaining name changes manually:
#     033E = 033E2
#     053E2 = 053D2
#     041E = 041E_2
#     075A2 = 075A1
tinmanSampleLUT <- sampleData %>% 
  filter(grepl("^TINMAN", sample_name)) %>% 
  select(sample_name) %>% 
  dplyr::rename(old_sample_name = sample_name) %>% 
  mutate(first = str_extract(old_sample_name, "^[:alpha:]+"),
         second = str_extract(old_sample_name, "[:digit:]+"),
         third = str_extract(old_sample_name, "[:alnum:]{2}$")) %>% 
  unite(new_sample_name, first, second, third) %>% 
  deframe()

sampleDataNewNames <- sampleData %>% 
  mutate(sample_name = case_when(sample_name %in% names(tinmanSampleLUT) ~ tinmanSampleLUT[sample_name],
                                 sample_name == "033E" ~ "033E_2",
                                 sample_name == "053E2" ~ "053D2",
                                 sample_name == "041E" ~ "041E_2",
                                 sample_name == "075A2" ~ "075A1",
                                 TRUE ~ sample_name)) %>% 
  as.data.frame()

# Add reads per sample from the phyloseq object
readsPerSample <- sample_sums(physeqRaw)

sampleDataCounts <- sampleDataNewNames %>% 
  mutate("readsPerSample" = readsPerSample[sample_name])

# Add unique_id to tell which samples belong to which patient (each patient
#   has two samples: a pre and post sample)
#   According to email, A/B/C are the first timepoints and D/E/F are second
#   except for samples 53, 61, 69, and 70, where A/B/C is on Abx, and D/E/F is
#   off abx long after treatment
# Also add a new "timepoint_condition" column that combines the pre_post and
#   neutropenic column. This will be used for the beta diversity plot.
sampleDataUniqueID <- sampleDataCounts %>%
  mutate(unique_id = str_extract(string = sample_name,
                                 pattern = "[:digit:]+"),
         timepoint_id = str_extract(string = sample_name,
                                    pattern = "[A-F]{1}.?[:digit:]{1}$"),
         timepoint_condition = paste(pre_post, neutropenic, sep = "_")) %>% 
  select(sample_name, unique_id, timepoint_id, pre_post, neutropenic, 
         timepoint_condition, everything())

sampleDataFinal <- sampleDataUniqueID

row.names(sampleDataFinal) <- sampleDataFinal$sample_name
```

```{r save-sampleDataFinal-for-ENA-use, eval=FALSE, include=FALSE}
saveRDS(sampleDataFinal, here("../data/RDataObjects/16S/sampleDataFinal.RDS"))
```

```{r physeqMerged, results='hide'}
physeqMerged <- merge_phyloseq(physeqRaw, sample_data(sampleDataFinal))
physeqMerged # 1823 taxa x 76 samples
```

```{r set-values-for-analyses}
# plotting colors
groupColors <- cbPaletteGrey[1:2]
names(groupColors) <- c("Control", "Neutropenic")

timepointColors <- c("PRE" = "blue", "POST" = "red")

tpCndColors <- cbPaletteGrey[5:8]
names(tpCndColors) <- unique(sampleDataFinal$timepoint_condition)

# Save columnNames to a reactiveValue as well so it can be accessed when 
#   running ADONIS calculations for beta diversity, and DESeq2 analysis
#   for biomarkers.
columnLUT <- reactiveValues("all_samples_by_cnd" = "neutropenic",
                            "all_samples_by_tp_cnd" = "timepoint_condition",
                            "pre_timepoint" = "neutropenic",
                            "post_timepoint" = "neutropenic",
                            "control_samples" = "pre_post",
                            "neutropenic_samples" = "pre_post")
```

### **QC**

<br>

Let's get an overview of the sample distribution by group and time point:

```{r sampleOverview}
sampleOverview <- sampleDataFinal %>% 
  tabyl(neutropenic, pre_post)

# options list dom = 't' means only the table shows (not the search or length menu)
DT::renderDataTable(sampleOverview, rownames = FALSE, options = list(dom = 't'))
```

<br>
<br>

<h4>**Read Count Distributions**</h4>

<br>

The samples will be compared in three different groupings. Therefore we will
first check for differences in the average reads/sample of each group, as any
significant differences could affect the outcome of the ecological measures.

```{r frequency-of-read-counts}
# make df of reads/sample
readsPerSampleDF <- sampleDataFinal %>% 
  select(sample_name, neutropenic, pre_post, readsPerSample) %>% 
  arrange(readsPerSample)

# apply  to samples so bar graph will show increasing reads/sample
sampleLevels <- pull(readsPerSampleDF, sample_name)
readsPerSampleDF$sample_name <- factor(readsPerSampleDF$sample_name,
                                       levels = sampleLevels)
```

```{r summaryReadDist}
summary(readsPerSampleDF$readsPerSample)
```

Reads/sample range from 1,521 to 45,885 with an average of 28,826.

```{r readDistPlot}
readDistPlot <- ggplot(readsPerSampleDF,
                       aes(x = sample_name, y = readsPerSample,
                           fill = neutropenic)) +
  scale_fill_manual(values = groupColors) +
  geom_bar(stat = "identity") +
  ggtitle("Read Count Distributions",
          subtitle = "Colored by Group") +
  theme_bw() +
  theme(axis.text.x = element_blank()) +
  guides(fill = guide_legend(title = "Group")) +
  facet_wrap(~ pre_post, scales = "free_x", nrow = 2, ncol = 1)

readDistPlot
```
```{r readDistBox}
readDistBox <- PrettyBox(df = sampleDataFinal,
                         xVar = "neutropenic", yVar = "readsPerSample",
                         statMethod = "wilcox.test",
                         plotTitle = "Avg. Reads Across Groups (Time Points Combined)",
                         colorVals = groupColors,
                         label_x = "Group", label_y = "Avg. Reads/Sample")

readDistBox
```

<br>

There is no significant difference in average reads per sample between the 
control and neutropenic sample groups when combined across pre/post time points.

```{r readDistBoxByTime}
readDistBoxByTime <- PrettyBox(df = sampleDataFinal,
                               xVar = "neutropenic", yVar = "readsPerSample",
                               statMethod = "wilcox.test",
                               plotTitle = "Avg. Reads Across Groups Within Time Point",
                               colorVals = groupColors,
                               label_x = "Time Point", label_y = "Avg. Reads/Sample",
                               facet_formula = "~ pre_post",
                               facet_rows = 2, facet_cols = 1)

readDistBoxByTime
```

<br>

Likewise, there is no significant difference in the average number of 
reads per sample between the control and neutropenic samples at pre or post
time points.

<br>

```{r readDistBoxByGroup}
readDistBoxByGroup <- PrettyBox(df = sampleDataFinal,
                               xVar = "pre_post", yVar = "readsPerSample",
                               statMethod = "wilcox.test",
                               plotTitle = "Avg. Reads Across Time Points Within Groups",
                               colorVals = timepointColors,
                               label_x = "Group", label_y = "Avg. Reads/Sample",
                               facet_formula = "~ neutropenic",
                               facet_rows = 2, facet_cols = 1)

readDistBoxByGroup
```

<br>

When looking at the time points within the individual groups, there is a significant
difference in average reads between the Pre and Post time points for the
control samples. This will have to be kept in mind if any significant differences
are found when comparing these two groups, as the communities may be sampled
unevenly.

<br>
<br>

---

<h4>**Sample & Taxa Filtering**</h4>

<br>

Non-bacterial taxa will be removed from the data set.

<br>

```{r filter-non-bacterial, results='hide'}
taxonomyTable <- as.data.frame(as(tax_table(physeqMerged), "matrix"))
uniqueTaxa <- map(taxonomyTable, unique)
taxTableMerged <- mutate(taxonomyTable, "KP" = paste(Kingdom, Phylum, sep = "_"))
unique(taxTableMerged$KP)

physeqBacteria <- subset_taxa(physeqMerged,
                              Kingdom == "Bacteria" &
                                Phylum != "Cyanobacteria" &
                                Phylum != "Chloroplast")
# any samples depleted?
any(sample_sums(physeqBacteria) == 0) # FALSE
physeqBacteria # 1809 x 76

sampleDataBacteria <- as(sample_data(physeqBacteria), "data.frame")
```

After dropping non-bacterial taxa, the total number of taxa are reduced from 
1,823 to 1,809.

```{r save-physeqBacteria, eval=FALSE, include=FALSE}
saveRDS(physeqBacteria, 
        here("../data/RDataObjects/16S/physeqObjects/physeqBacteria.RDS"))
```


Now checking the prevalence and abundance of different bacterial families present.
It may be useful to identify and remove low prevalence/low abundance taxa from
the data set to help with downstream analysis, specifically when looking for
biomarkers.

```{r phylaPrevalence}
phylaPrevalence <- TaxRankPrevalence(physeqBacteria, "Phylum")

phylaPrevalencePlot <- ggplot(phylaPrevalence$prevalence_df,
                              aes(x = TotalAbundance,
                                  y = Prevalence/nsamples(physeqBacteria),
                                  color = Family)) +
  geom_hline(yintercept = 1/nsamples(physeqBacteria), alpha = 0.5, 
             linetype = 2) +
  geom_point(size = 3, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") +
  ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~ Phylum) +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  ggtitle("Phylum Prevalence in All Samples",
          subtitle = "Colored by Familiy")
phylaPrevalencePlot
```

```{r physeqSubsetList}
# subset physeqBacteria by:
#   all (physeqBacteria)
#   pre time point (pre_post == "PRE")
#   post time point (pre_post == "POST)
#   control group only (neutropenic == "Control")
#   neutropenic group only (neutropenic == "Neutropenic")

subsetNames <- c("pre_timepoint",
                 "post_timepoint",
                 "control_samples",
                 "neutropenic_samples")

# What column to use to make each subset in subsetName:
columnNames <- c(rep("pre_post", 2), rep("neutropenic", 2))
names(columnNames) <- subsetNames

# What labels meet each condition:
conditionValues <- c("PRE",
                     "POST",
                     "Control",
                     "Neutropenic")
names(conditionValues) <- subsetNames

# Initialize list to hold the physeqSubsets
physeqSubsetList <- vector(mode = "list", length = length(subsetNames))

for (i in 1:length(subsetNames)) {

  currentSubset <- subsetNames[i]
  currentColumn <- columnNames[i]
  currentCondition <- conditionValues[i]
  
  keepSamples <- sampleDataBacteria %>% 
    filter(!!sym(currentColumn) == currentCondition) %>% 
    pull(sample_name)
  
  psSubset <- prune_samples(keepSamples, physeqBacteria) %>% 
    RemoveMissingTaxa()
  
  physeqSubsetList[[i]] <- psSubset
  names(physeqSubsetList)[i] <- currentSubset
}

# Add all_samples as physeqBacteria
physeqSubsetList <- c("all_samples_by_cnd" = physeqBacteria, 
                      "all_samples_by_tp_cnd" = physeqBacteria,
                      physeqSubsetList)
```

<br>
<br>

### **Community Composition**

<br>

Community composition plots can be used to examine the overall taxon representation
across samples.

```{r comm-comp}

#----- UI -----#

fluidRow(
  
  column(5,
         
         wellPanel(
           # select rank to display
           selectInput(inputId = "communityTaxRank", label = "Taxonomic Rank:",
                       choices = c("Kingdom", "Phylum", "Class", "Order",
                                   "Family", "Genus", "Species"),
                       selected = "Phylum"),
           
           # select abudnance filter
           sliderInput(inputId = "abdFilter", label = "Abundance Filter",
                       min = 0, max = 1, step = 0.05, value = 0),
           # update
           actionButton(inputId = "updateCommunity", 
                        label = "Update Community Plot",
                        icon = icon("fas fa-sync-alt"))
         ))
  
)

#----- React -----#

communityCompXLabs <- sampleDataFinal %>% 
  select(sample_name, unique_id) %>%
  mutate(unique_id = str_remove(string = unique_id, "^0+")) %>% 
  deframe()

GetCommunityRank <- reactive({
  
  rank <- input$communityTaxRank
  
  if (is.null(input$communityTaxRank)) {
    rank <- "Phylum"
  }

  return(rank)
  
})

CalculateAbundance <- eventReactive(input$updateCommunity, {
  
  withProgress(message = "Calculating Abundance...", value = 1, {
    
    abdDF <- MakeAbundanceDF(physeq = physeqBacteria,
                             taxRank = isolate(GetCommunityRank()),
                             abundanceFilter = isolate(input$abdFilter),
                             pruneMissing = FALSE)
    
    # apply levels to the Sample column of abdDF
    sampleLevels <- abdDF %>% 
      select(Sample, unique_id, pre_post) %>% 
      distinct() %>% 
      arrange(unique_id, pre_post) %>% 
      pull(Sample)
    
    abdDF$Sample <- factor(abdDF$Sample, levels = sampleLevels)
    
    if (dim(abdDF)[1] == 0) {
      stop("No taxa meet this filtering criteria. Try lowering the Abundance Filter option.",
           call. = FALSE)
    }
    
    if (isolate(GetCommunityRank()) == "Species") {
      abdDF <- mutate(abdDF, "GenusSpecies" = paste(Genus, Species, sep = " "))
    }
    
    return(abdDF)
    
  })
  
}, ignoreNULL = FALSE)

PlotCommunityCompositionModified <- function(abdDF, taxRank = "Phylum",
                                             facetFormula = NULL,
                                             facetCol = NULL, facetRow = NULL) {
  basePlot <- ggplot(abdDF,
                     aes_string(x = "Sample", y = "Abundance",
                                fill = taxRank)) +
    geom_bar(stat = "identity", width = 1, color = "grey14") +
    xlab("Subject") +
    ylab("Relative Abundance") +
    scale_x_discrete(labels = communityCompXLabs) +
    theme(axis.text.x = element_text(size = 12, angle = -90),
          plot.title = element_text(hjust = 0.5),
          plot.subtitle = element_text(hjust = 0.5))
  # are there facets??
  if (!(is.null(facetFormula))) {
    formula <- as.formula(facetFormula)
    facetPlot <- basePlot +
      facet_wrap(formula, scales = "free", nrow = facetRow, ncol = facetCol) +
      theme(panel.spacing.y = unit(1, "cm", data = NULL))
    return(facetPlot)
  } else {
    return(basePlot)
  }
  
}

renderPlotly({
  
  withProgress(message = "Loading plot...", value = 1, {
    
    compPlot <- PlotCommunityCompositionModified(abdDF = CalculateAbundance(),
                                                 taxRank = ifelse(
                                                   isolate(GetCommunityRank()) == "Species",
                                                   yes = "GenusSpecies",
                                                   no = isolate(GetCommunityRank())),
                                                 facetFormula = "pre_post ~ neutropenic",
                                                 facetCol = 2, facetRow = 2)
    ggplotly(compPlot, height = 800, width = 1000)
  
  })

})
```

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

```{r save-static-community-composition-plots, eval=FALSE, include=FALSE}
rankList <- c("Phylum", "Class", "Order", "Family", "Genus", "Species")
names(rankList) <- rankList

# Get sample order (place 102 b/w 50 and 53)
sampleLevels <- names(communityCompXLabs)[order(names(communityCompXLabs))]
moveSamples <- sampleLevels[grepl("102", sampleLevels)]
sampleLevelsShort <- sampleLevels[sampleLevels != moveSamples]
sampleLevelsFinal <- c(sampleLevelsShort[1:38],
                       moveSamples,
                       sampleLevelsShort[39:74])

# Make abundance DF for plotting at each tax rank
abdDFList <- map(.x = rankList,
                 .f = MakeAbundanceDF,
                 physeq = physeqBacteria,
                 abundanceFilter = 0,
                 pruneMissing = FALSE)

# For species abundance DF, make Genus Species column to account for the many
#   NAs in the species column
abdDFList$Species <- mutate(abdDFList$Species, 
                            "Species" = paste(Genus, Species, sep = " "))

communityPlotList <- map2(.x = abdDFList,
                          .y = rankList,
                          .f = ~ PlotCommunityCompositionModified(abdDF = .x,
                                                                  taxRank = .y,
                                                                  facetFormula = "pre_post ~ neutropenic",
                                                                  facetCol = 2,
                                                                  facetRow = 2))

communityPlotList

# Apply sampleLevelsFinal to the Phylum and Class level community plots
abdDFPhylum <- abdDFList$Phylum
abdDFPhylum$Sample <- factor(abdDFPhylum$Sample, levels = sampleLevelsFinal)
communityPhylumLevel <- PlotCommunityCompositionModified(abdDF = abdDFPhylum,
                                                         taxRank = "Phylum",
                                                         facetFormula = "pre_post ~ neutropenic",
                                                         facetCol = 2,
                                                         facetRow = 2)
abdDFClass <- abdDFList$Class
abdDFClass$Sample <- factor(abdDFClass$Sample, levels = sampleLevelsFinal)
communityClassLevel <- PlotCommunityCompositionModified(abdDF = abdDFClass,
                                                        taxRank = "Class",
                                                        facetFormula = "pre_post ~ neutropenic",
                                                        facetCol = 2,
                                                        facetRow = 2)

# Save Phylum and Class for now because the legends get huge otherwise
communityPlotsPath <- paste0(figPath, "community_composition")

if (!file.exists(communityPlotsPath)) {
  dir.create(communityPlotsPath)
}

communityPlotNames <- paste0(communityPlotsPath, "/", c("Phylum", "Class"),
                             "-community-composition.pdf")
walk2(.x = list(communityPhylumLevel, communityClassLevel),
      .y = here(communityPlotNames),
      .f = ~ ggsave(filename = .y, plot = .x, width = 11, height = 8.5,
                    units = "in"))
```


### **Alpha Diversity**

<br>

Alpha diversity is concerned with describing the community within a sample. 
It is a standard tool used to calculate ecological indices which give the number 
of taxa present in a sample and the relationships between relative abundance and 
how evenly taxa are distributed. These indices provide useful summary 
information about the community structure of a given sample.

We are primarily interested in richness (the number of taxa per sample) and 
Shannon diversity (a description of the richness and evenness of a sample's 
community). An evenness of 0 indicates that a community is dominated by one or 
a few taxa, but an evenness of 1 means that species are evenly distributed. 
Diversity increases as richness and evenness increase.

```{r alphaDiv}
alphaIdx <- c("Richness" = "observed",
              "Peilou's Evenness" = "evenness_pielou",
              "Shannon Diversity" = "diversity_shannon")
alphaDiv <- microbiome::alpha(physeqBacteria,
                              index = alphaIdx)

sampleDataBacteria <- base::merge(alphaDiv, sampleDataBacteria, by = "row.names") %>% 
  column_to_rownames(var = "Row.names")
```

<br>
<br>

```{r richness-stats, eval=FALSE, include=FALSE}
write.table(sampleDataBacteria,
            file = here("../analysis/results/16S/16S_metadataWithAlphaDiv.txt"),
            quote = FALSE, sep = "\t", row.names = FALSE)

richnessStats <- sampleDataBacteria %>% 
  group_by(timepoint_condition) %>% 
  summarize("avg_richness" = mean(observed),
            "med_richness" = median(observed))

write.table(richnessStats, 
            file = here("../analysis/results/16S/richnessStats.txt"),
            quote = FALSE, sep = "\t", row.names = FALSE)
```

```{r PlotAlphaModified}
# Modified PrettyBox() function to add option for connecting samples by unique_id
#   and setting limits on the y-axes

PlotAlphaModified <- function(df, xVar, yVar, statMethod, plotTitle, yLims,
                              addLines = FALSE,
                              colorVals = NULL, label_x = NULL, label_y = NULL,
                              legendPos = "none", facet_formula = NULL,
                              facet_rows = NULL, facet_cols = NULL) {
  
  basePlot <- ggplot(data = df,
                     aes_string(x = xVar, y = yVar, fill = xVar)) +
    geom_boxplot(outlier.shape = NA) +
    scale_fill_manual(values = colorVals) +
    labs(x = label_x, y = label_y, title = plotTitle) +
    scale_y_continuous(limits = yLims) +
    theme_pubr() +
    theme(axis.title = element_text(size = 14),
          axis.text = element_text(size = 12),
          plot.title = element_text(size = 14, hjust = 0.5),
          strip.text = element_text(size = 12),
          legend.title = element_blank(), 
          legend.position = legendPos,
          legend.text = element_text(size = 12)) +
    stat_compare_means(method = statMethod, label.x.npc = 0)
  
  if (addLines) {
    basePlot <- basePlot +
      geom_point(size = 1.5) +
      geom_line(aes(group = unique_id))
  } else {
    basePlot <- basePlot +
      geom_jitter(width = 0.2, size = 1.5)
  }
  
  if (!is.null(facet_formula)) {
    basePlot <- basePlot +
      facet_wrap(as.formula(facet_formula), nrow = facet_rows, ncol = facet_cols,
                 scales = "free")
  }
  
  return(basePlot)
  
}

PlotAlphaViolin <- function(df, xVar, yVar, statMethod, plotTitle, yLims,
                            addLines = FALSE,
                            colorVals = NULL, label_x = NULL, label_y = NULL,
                            legendPos = "none", facet_formula = NULL,
                            facet_rows = NULL, facet_cols = NULL) {
  basePlot <- ggplot(data = df,
                     aes_string(x = xVar, y = yVar, fill = xVar)) +
    geom_violin() +
    geom_boxplot(width = 0.1, color = "black", outlier.shape = NA) +
    #geom_jitter(width = 0.05, color = "black") +
    scale_fill_manual(values = colorVals) +
    labs(x = label_x, y = label_y, title = plotTitle) +
    scale_y_continuous(limits = yLims) +
    theme_pubr() +
    theme(axis.title = element_text(size = 14),
          axis.text = element_text(size = 12),
          plot.title = element_text(size = 14, hjust = 0.5),
          strip.text = element_text(size = 12),
          legend.title = element_blank(), 
          legend.position = legendPos,
          legend.text = element_text(size = 12)) +
    stat_compare_means(method = statMethod, label.x.npc = 0)
  
  if (addLines) {
    basePlot <- basePlot +
      geom_point() +
      geom_line(aes(group = unique_id))
  } else {
    basePlot <- basePlot +
      geom_jitter(width = 0.05, color = "black")
  }
  
  if (!is.null(facet_formula)) {
    basePlot <- basePlot +
      facet_wrap(as.formula(facet_formula), nrow = facet_rows, ncol = facet_cols,
                 scales = "free")
  }
  
  return(basePlot)
  
}

alphaPlotsPath <- paste0(figPath, "alpha_diversity")
```

```{r alphaCombinedTP, fig.width=9, fig.height=4}
#----- Combined Time Points -----#

alphaCombinedTP <- pmap(.l = list(yVar = alphaIdx,
                                  plotTitle = names(alphaIdx),
                                  yLims = list("Richness" = c(0, 280),
                                               "Peilou's Evenness" = c(0, 1),
                                               "Shannon Diversity" = c(0, 4.5))),
                        .f = PlotAlphaModified,
                        df = sampleDataBacteria,
                        xVar = "neutropenic",
                        statMethod = "wilcox.test",
                        colorVals = groupColors)

alphaCombinedTPArr <- annotate_figure(ggarrange(plotlist = alphaCombinedTP,
                                                nrow = 2, ncol = 2),
                                      top = text_grob("Alpha Diversity (Time Points Combined)",
                                                      size = 14, face = "bold"))
alphaCombinedTPArr
```

```{r save-alphaCombinedTPArr, eval=FALSE}
ggsave(paste0(alphaPlotsPath, "/alpha-combined-timepoints.pdf"),
       alphaCombinedTPArr, width = 11, height = 8.5, units = "in")
```

<br>
<br>

---

<br>
<br>

```{r alphaSeparateTP, fig.width=9, fig.height=6}

#----- Separate Time Points -----#

alphaSeparateTP <- pmap(.l = list(yVar = alphaIdx,
                                  plotTitle = names(alphaIdx),
                                  yLims = list("Richness" = c(0, 280),
                                               "Peilou's Evenness" = c(0, 1),
                                               "Shannon Diversity" = c(0, 4.5))),
                        .f = PlotAlphaModified,
                        df = sampleDataBacteria,
                        xVar = "neutropenic",
                        statMethod = "wilcox.test",
                        colorVals = groupColors,
                        facet_formula = "~ pre_post",
                        facet_rows = 2, facet_col = 1)

alphaSeparateTPArr <- annotate_figure(ggarrange(plotlist = alphaSeparateTP,
                                                nrow = 1, ncol = 3),
                                      top = text_grob("Alpha Diversity at Pre and Post Time Points",
                                                      size = 14, face = "bold"))
alphaSeparateTPArr
```

```{r save-alphaSeparateTPArr, include=FALSE, eval=FALSE}
ggsave(paste0(alphaPlotsPath, "/alpha-separate-timepoints.pdf"),
       alphaSeparateTPArr, width = 11, height = 8.5, units = "in")
```

<br>
<br>

---

<br>
<br>

```{r alphaPrePost, fig.width=9, fig.height=6}
alphaPrePost <- pmap(.l = list(yVar = alphaIdx,
                               plotTitle = names(alphaIdx),
                               yLims = list("Richness" = c(0, 280),
                                               "Peilou's Evenness" = c(0, 1),
                                               "Shannon Diversity" = c(0, 4.5))),
                     .f = PlotAlphaModified,
                     addLines = TRUE,
                     df = sampleDataBacteria,
                     xVar = "pre_post",
                     statMethod = "wilcox.test",
                     colorVals = timepointColors,
                     facet_formula = "~ neutropenic",
                     facet_rows = 2, facet_col = 1)

alphaPrePostArr <- annotate_figure(ggarrange(plotlist = alphaPrePost,
                                             nrow = 1, ncol = 3),
                                   top = text_grob("Alpha Diversity of Pre vs Post Time Points Within Groups",
                                                   size = 14, face = "bold"))
alphaPrePostArr
```

```{r save-alphaPrePostArr, include=FALSE, eval=FALSE}
ggsave(paste0(alphaPlotsPath, "/alpha-pre-post.pdf"),
       alphaPrePostArr, width = 11, height = 8.5, units = "in")
```

<br>
<br>

Richness is significantly higher in the "pre" time point for both the control and
neutropenic groups. This may be expected following antibiotics.

<br>
<br>

--

<br>
<br>

<h4> Change in Alpha Diversity Over Time: Control vs. Neutropenic </h4>

<br>
<br>

Here we are looking to see if the change in alpha diversity from the Pre to Post
timepoint is significantly different in the control samples versus the 
neutropenic samples. In this plot, each point represents the change in richness
for a given sample.

```{r calculate-change-in-alpha-diversity}
deltaMetrics <- list("Richness" = "observed",
                     "Shannon Diversity" = "diversity_shannon")

MakeDeltaDF <- function(metric) {
  
  newColName <- paste0("delta_", metric)
  
  deltaDF <- sampleDataBacteria %>%
    select(unique_id, !!sym(metric), pre_post, neutropenic) %>%
    pivot_wider(id_cols = c(unique_id, neutropenic),
                names_from = pre_post,
                values_from = !!sym(metric)) %>%
    mutate(!!newColName := abs(POST - PRE))
  
  return(deltaDF)
}

deltaAlphaDFList <- map(.x = deltaMetrics, .f = MakeDeltaDF)

deltaAlphaPlots <- pmap(.l = list(df = deltaAlphaDFList,
                                  yVar = list("Richness" = "delta_observed",
                                              "Shannon Diversity" = "delta_diversity_shannon"),
                                  label_y = paste("Change in",
                                                  names(deltaAlphaDFList)),
                                  plotTitle = paste("Change in",
                                                    names(deltaAlphaDFList),
                                                    "Over Time")),
                        .f = PrettyBox,
                        xVar = "neutropenic",
                        statMethod = "wilcox.test",
                        colorVals = groupColors)

deltaAlphaPlotsArr <- ggarrange(plotlist = deltaAlphaPlots,
                                nrow = 1, ncol = 2)
deltaAlphaPlotsArr
```

<br>
<br>

The average change in richness and Shannon diversity over time does not vary
significantly between control and neutropenic samples.

<br>
<br>

### **Beta Diversity**

<br>

While alpha diversity was concerned with describing the community within a 
sample, beta diversity compares the communities between samples. Beta diversity 
is quantified through "association" coefficients - similarity or distance 
measures. It describes how similar/dissimilar samples' communities are to one 
another. We view these distances with ordination plots such as PCoA.

A common distance measure is the UniFrac distance, which is a measure that 
considers the phylogenetic relationships between bacterial taxa. A community of 
more closely-related taxa is less diverse than a community of distantly-related 
taxa. The weighted UniFrac measure additionally accounting for the relative 
abundance of each taxa. 

<br>
<br>

```{r beta-diversity-ui}
fluidRow(
  column(5,
         wellPanel(
           # select group to plot
           selectInput(inputId = "betaGroup",
                       label = "Select Group",
                       choices = c("All (by condition)" = "all_samples_by_cnd",
                                   "All (by timepoint & condition)" = "all_samples_by_tp_cnd",
                                   "PRE Time Point" = "pre_timepoint",
                                   "POST Time Point" = "post_timepoint",
                                   "Control" = "control_samples",
                                   "Neutropenic" = "neutropenic_samples"),
                       selected = "All (by condition)"),
           # select distance measure
           selectInput(inputId = "betaDistance",
                       label = "Distance Measure",
                       choices = c("UniFrac" = "unifrac",
                                   "Weighted UniFrac" = "wunifrac"),
                       selected = "UniFrac"),
           # show ellipses?
           checkboxInput(inputId = "betaEllipsesBool",
                         label = "Show Confidence Ellipses",
                         value = FALSE),
           # show lines connecting pre/post samples
           checkboxInput(inputId = "betaShowLines",
                         label = "Show Lines Connecting Subject Samples",
                         value = FALSE),
           # update
           actionButton(inputId = "betaUpdate",
                        label = "Update Beta Diversity Plots",
                        icon = icon("fas fa-sync-alt"))
         ))
)
```

```{r beta-diversity-reactive}

#----- Handle Start Up NULLS -----#

SelectedPhyseq <- reactive({
  selectedPhyseq <- input$betaGroup
  if (is.null(selectedPhyseq)) {
    selectedPhyseq <- "all_samples_by_cnd"
  }
  return(selectedPhyseq)
})

SelectedDistance <- reactive({
  selectedDistance <- input$betaDistance
  if (is.null(selectedDistance)) {
    selectedDistance <- "unifrac"
  }
  return(selectedDistance)
})


#----- Calculate Ordination -----#

CalculateOrdination <- eventReactive(input$betaUpdate, {
  withProgress(message = "Calculating Ordination...", value = 1, {
    set.seed(11643764)
    ordinationObj <- ordinate(physeqSubsetList[[SelectedPhyseq()]],
                              method = "PCoA",
                              distance = SelectedDistance())
  })
}, ignoreNULL = FALSE)

#----- Run ADONIS for Significance -----#

AdonisPVal <- eventReactive(input$betaUpdate, {
  set.seed(11643764)
  adonisRes <- RunAdonis(physeqSubsetList[[SelectedPhyseq()]],
                         category = columnLUT[[SelectedPhyseq()]],
                         distance = SelectedDistance())
  pVal <- adonisRes$aov.tab$`Pr(>F)`[1]
}, ignoreNULL = FALSE)


#----- Render the Ordination Plot -----#

observe({print(SelectedPhyseq())})

MakeOrdPlot <- eventReactive(input$betaUpdate, {
  
  colorOptions <- list("all_samples_by_cnd" = groupColors,
                       "all_samples_by_tp_cnd" = tpCndColors,
                       "pre_timepoint" = groupColors,
                       "post_timepoint" = groupColors,
                       "control_samples" = timepointColors,
                       "neutropenic_samples" = timepointColors)
  
  titleOptions <- c("all_samples_by_cnd" = "All Samples",
                    "all_samples_by_tp_cnd" = "All Samples",
                    "pre_timepoint" = "Pre Time Point Samples",
                    "post_timepoint" = "Post Time Point Samples",
                    "control_samples" = "Control Samples",
                    "neutropenic_samples" = "Neutropenic Samples")
  
  ordPlot <- plot_ordination(physeq = physeqSubsetList[[SelectedPhyseq()]],
                             ordination = CalculateOrdination(),
                             color = columnLUT[[SelectedPhyseq()]]) +
    scale_color_manual(values = colorOptions[[SelectedPhyseq()]]) +
    theme_bw() +
    geom_point(size = 3.5) +
    ggtitle(paste(toupper(SelectedDistance()),
                  "Distance of", 
                  titleOptions[[SelectedPhyseq()]]),
            subtitle = paste("ADONIS p-val:", AdonisPVal())) +
    theme(plot.title = element_text(hjust = 0.5, size = 16),
          plot.subtitle = element_text(hjust = 0.5, size = 14),
          legend.title = element_text(size = 16),
          legend.text = element_text(size = 14))
  
  # Show ellipses?
  if (input$betaEllipsesBool) {
    ordPlot <- ordPlot + stat_ellipse(type = "norm")
  }
  
  # Show lines?
  if (input$betaShowLines) {
    ordPlot <- ordPlot +
      geom_line(aes_string(group = "unique_id"))
  }
  
  return(ordPlot)
  
}, ignoreNULL = FALSE)


renderPlot({
  
  MakeOrdPlot()
  
}, height = 600, width = 900)
```

<br>
<br>

According to ADONIS, the pre_post variable is significant to the clustering of
control samples when using both UniFrac and weighted UniFrac. It is also 
significant to the grouping of neutropenic samples when using the UniFrac
metric only.

<br>
<br>

```{r save-static-ordination-plots, eval=FALSE, include=FALSE}

# Make static ordination plots for all ps objects in physeqSubsetList

#----- Calculate Ordinations -----#

set.seed(11643764)
unweightedOrdList <- map(.x = physeqSubsetList,
                         .f = ~ ordinate(.x, method = "PCoA", "unifrac"))

set.seed(11643764)
weightedOrdList <- map(.x = physeqSubsetList,
                       .f = ~ ordinate(.x, method = "PCoA", "wunifrac"))


#----- Run ADONIS for Significance -----#

variableColNames <- list("all_samples_by_cnd" = "neutropenic",
                         "all_samples_by_tp_cnd" = "timepoint_condition",
                         "pre_timepoint" = "neutropenic",
                         "post_timepoint" = "neutropenic",
                         "control_samples" = "pre_post",
                         "neutropenic_samples" = "pre_post")

set.seed(11643764)
unweightedAdonisResList <- map2(.x = physeqSubsetList,
                                .y = variableColNames,
                                .f = ~ RunAdonis(physeqObj = .x,
                                                 category = .y,
                                                 distance = "unifrac"))
unweightedPVals <- map(.x = unweightedAdonisResList,
                       .f = ~ .x$aov.tab$`Pr(>F)`[1])

set.seed(11643764)
weightedAdonisResList <- map2(.x = physeqSubsetList,
                              .y = variableColNames,
                              .f = ~ RunAdonis(physeqObj = .x,
                                               category = .y,
                                               distance = "wunifrac"))
weightedPVals <- map(.x = weightedAdonisResList,
                     .f = ~ .x$aov.tab$`Pr(>F)`[1])

#----- Draw Ordination Plots (unweighted & weighted) with Confidence Ellipses -----#

colorByVar <- c("all_samples_by_cnd" = "neutropenic",
                "all_samples_by_tp_cnd" = "timepoint_condition",
                "pre_timepoint" = "neutropenic",
                "post_timepoint" = "neutropenic",
                "control_samples" = "pre_post",
                "neutropenic_samples" = "pre_post")

colorValues <- list("all_samples_by_cnd" = groupColors,
                 "all_samples_by_tp_cnd" = tpCndColors,
                 "pre_timepoint" = groupColors,
                 "post_timepoint" = groupColors,
                 "control_samples" = timepointColors,
                 "neutropenic_samples" = timepointColors)
  
titleOptions <- c("all_samples_by_cnd" = "All Samples",
                  "all_samples_by_tp_cnd" = "All Samples",
                  "pre_timepoint" = "Pre Time Point Samples",
                  "post_timepoint" = "Post Time Point Samples",
                  "control_samples" = "Control Samples",
                  "neutropenic_samples" = "Neutropenic Samples")

PlotOrdinationsModified <- function(physeqObj, ordinationObj, colorBy,
                                    colorValues, distanceMetric,
                                    titleString, adonisP) {
  
  ordPlot <- plot_ordination(physeq = physeqObj,
                             ordination = ordinationObj,
                             color = colorBy) +
    scale_color_manual(values = colorValues) +
    theme_bw() +
    geom_point(size = 3.5) +
    ggtitle(paste(toupper(distanceMetric),
                  "Distance of", 
                  titleString),
            subtitle = paste("ADONIS p-val:", adonisP)) +
    theme(plot.title = element_text(hjust = 0.5, size = 16),
          plot.subtitle = element_text(hjust = 0.5, size = 14),
          legend.title = element_text(size = 16),
          legend.text = element_text(size = 14)) + 
    stat_ellipse(type = "norm")
}

unweightedOrdPlots <- pmap(.l = list(physeqObj = physeqSubsetList,
                                     ordinationObj = unweightedOrdList,
                                     colorBy = colorByVar,
                                     colorValues = colorValues,
                                     titleString = titleOptions,
                                     adonisP = unweightedPVals),
                           .f = PlotOrdinationsModified,
                           distanceMetric = "unifrac")

weightedOrdPlots <- pmap(.l = list(physeqObj = physeqSubsetList,
                                   ordinationObj = weightedOrdList,
                                   colorBy = colorByVar,
                                   colorValues = colorValues,
                                   titleString = titleOptions,
                                   adonisP = weightedPVals),
                         .f = PlotOrdinationsModified,
                         distanceMetric = "wunifrac")

#----- Save Plots -----#

betaPlotsPath <- paste0(figPath, "beta_diversity")

if (!file.exists(betaPlotPaths)) {
  dir.create(betaPlotsPath)
}

unweightedPlotFileNames <- paste0(betaPlotsPath, 
                                  "/unweighted_UniFrac_ordination_of_",
                                  names(unweightedOrdPlots),
                                  ".pdf")

walk2(.x = unweightedPlotFileNames,
      .y = unweightedOrdPlots,
      .f = ~ ggsave(filename = .x, plot = .y, width = 11, height = 8.5,
                    units = "in"))

weightedPlotFileNames <- paste0(betaPlotsPath,
                                "/weighted_UniFrac_ordination_of_",
                                names(weightedOrdPlots),
                                ".pdf")

walk2(.x = weightedPlotFileNames,
      .y = weightedOrdPlots,
      .f = ~ ggsave(filename = .x, plot = .y, width = 11, height = 8.5,
                    units = "in"))

```

### **Biomarker Analysis with DESeq2**

<br>

We will use DESeq2 to investigate potential biomarkers between groups. This 
information may be referenced against the community composition plots to
identify broad taxonomic groups that may identify a subset of samples.

<br>
<br>

```{r GenerateDESeqResultsModified}
GenerateDESeqResultsModified <- function(physeq, variable, numerator, denominator) {
  
  # Returns DESeq Results as Formal Class "DESeqResults"
  # Create formula from string variable
  #formula <- as.formula(paste("~ pre_post +", variable, sep = " "))
  formula <- as.formula(paste("~", variable, sep = " "))
  # Convert to deseq data set object
  ds <- phyloseq_to_deseq2(physeq, design = formula)
  # work-around
  counts <- counts(ds)
  geoMeans <- apply(counts, 1,
                      function(row) {if (all(row == 0)) 0 else exp(mean(log(row[row != 0])))})
  dds <- estimateSizeFactors(ds, geoMeans = geoMeans)
  # Run analysis
  ddsAnalysis <- DESeq(dds, test = "Wald", fitType = "local", betaPrior = FALSE)
  # Extract and format results
  ddsResults <- results(ddsAnalysis,
                        contrast = c(variable, numerator, denominator)) 
}
```

```{r set-reactiveValues-for-biomarker-analyses}

numeratorLUT <- reactiveValues("all_samples_by_cnd" = "Control",
                               "pre_timepoint" = "Control",
                               "post_timepoint" = "Control",
                               "control_samples" = "PRE",
                               "neutropenic_samples" = "PRE")

denominatorLUT <- reactiveValues("all_samples_by_cnd" = "Neutropenic",
                                 "pre_timepoint" = "Neutropenic",
                                 "post_timepoint" = "Neutropenic",
                                 "control_samples" = "POST",
                                 "neutropenic_samples" = "POST")
```

```{r biomarker-ui}
fluidRow(
  column(5,
         wellPanel(
           # select group to plot
           selectInput(inputId = "biomarkerComparison",
                       label = "Select Group",
                       choices = c("All" = "all_samples_by_cnd",
                                   "PRE Time Point" = "pre_timepoint",
                                   "POST Time Point" = "post_timepoint",
                                   "Control" = "control_samples",
                                   "Neutropenic" = "neutropenic_samples"),
                       selected = "All"),
           # update
           actionButton(inputId = "updateBiomarkers",
                        label = "Run DESeq2",
                        icon = icon("fas fa-sync-alt"))
         ))
)
```

```{r biomarker-reactive}
# Set initial value of biomarkerComparison
BiomarkerComparison <- reactive({
  biomarkerComparison <- input$biomarkerComparison
  
  if (is.null(biomarkerComparison)) {
    biomarkerComparison <- "all_samples_by_cnd"
  }
  
  return(biomarkerComparison)
})

RunDESeq <- eventReactive(input$updateBiomarkers, {
  
  withProgress(message = "Conducting DESeq2 Analysis...", value = 1, {
    
    colName <- columnLUT[[BiomarkerComparison()]]
    numeratorVar <- numeratorLUT[[BiomarkerComparison()]]
    denominatorVar <- denominatorLUT[[BiomarkerComparison()]]
    
    deseqRes <- GenerateDESeqResultsModified(physeq = physeqSubsetList[[BiomarkerComparison()]],
                                             variable = colName,
                                             numerator = numeratorVar,
                                             denominator = denominatorVar)
  })
  
}, ignoreNULL = FALSE)

DESeqRes <- eventReactive(input$updateBiomarkers, {
  
  deseqResTable <- GenerateDESeqResultsTable(physeq = physeqSubsetList[[BiomarkerComparison()]],
                                             sigThreshold = 0.05,
                                             ddsResults = RunDESeq())
}, ignoreNULL = FALSE)
```

<br>

```{r biomarker-plot}
renderPlotly({
  volcano <- PlotStaticVolcano(physeq = physeqSubsetList[[BiomarkerComparison()]],
                               resultsDataTable = DESeqRes(),
                               sigThreshold = 0.05,
                               plotTitle = paste("Differentially Abundant Taxa\n",
                                                 denominatorLUT[[isolate(BiomarkerComparison())]],
                                                 "(-L2FC) vs",
                                                 numeratorLUT[[isolate(BiomarkerComparison())]],
                                                 "(+ L2FC)\n SILVA Data"))
  
  ggplotly(volcano, tooltip = c("Phylum", "Family", "Genus", "Species",
                                "log2FoldChange", "baseMean"),
           height = 550, width = 900)
})
```

<br>
<br>
<br>
<br>

```{r save-static-biomarker-plots, eval=FALSE, include=FALSE}
numeratorLUT <- c("all_samples_by_cnd" = "Control",
                  "pre_timepoint" = "Control",
                  "post_timepoint" = "Control",
                  "control_samples" = "PRE",
                  "neutropenic_samples" = "PRE")

denominatorLUT <- c("all_samples_by_cnd" = "Neutropenic",
                    "pre_timepoint" = "Neutropenic",
                    "post_timepoint" = "Neutropenic",
                    "control_samples" = "POST",
                    "neutropenic_samples" = "POST")

comparisons <- names(numeratorLUT)
biomarkerPhyseqList <- physeqSubsetList[names(physeqSubsetList) %in% comparisons]

biomarkerColumns <- c("all_samples_by_cnd" = "neutropenic",
                      "pre_timepoint" = "neutropenic",
                      "post_timepoint" = "neutropenic",
                      "control_samples" = "pre_post",
                      "neutropenic_samples" = "pre_post")

formulaStrList <- c("all_samples_by_cnd" = "~ neutropenic",
                    "pre_timepoint" = "~ neutropenic",
                    "post_timepoint" = "~ neutropenic",
                    "control_samples" = "~ pre_post",
                    "neutropenic_samples" = "~ pre_post")

GenerateDESeqResultsModified <- function(physeq, variable, formulaStr,
                                         numerator, denominator) {
  
  # Returns DESeq Results as Formal Class "DESeqResults"
  # Create formula from string variable
  #formula <- as.formula(paste("~", variable, sep = " "))
  formula <- as.formula(formulaStr)
  # Convert to deseq data set object
  ds <- phyloseq_to_deseq2(physeq, design = formula)
  # work-around
  counts <- counts(ds)
  geoMeans <- apply(counts, 1,
                      function(row) {if (all(row == 0)) 0 else exp(mean(log(row[row != 0])))})
  dds <- estimateSizeFactors(ds, geoMeans = geoMeans)
  # Run analysis
  ddsAnalysis <- DESeq(dds, test = "Wald", fitType = "local", betaPrior = FALSE)
  # Extract and format results
  ddsResults <- results(ddsAnalysis,
                        contrast = c(variable, numerator, denominator)) 
}

deseqResList <- pmap(.l = list(physeq = biomarkerPhyseqList,
                               variable = biomarkerColumns,
                               numerator = numeratorLUT,
                               denominator = denominatorLUT,
                               formulaStr = formulaStrList),
                     .f = GenerateDESeqResultsModified)


deseqResTableList <- map2(.x = biomarkerPhyseqList,
                          .y = deseqResList,
                          .f = ~ GenerateDESeqResultsTable(physeq = .x,
                                                           sigThreshold = 0.05,
                                                           ddsResults = .y))

volcanoTitleStrings <- paste(names(biomarkerPhyseqList), "\n",
                             "Differentially Abundant Taxa\n",
                             denominatorLUT,
                             "(-L2FC) vs",
                             numeratorLUT,
                             "(+L2FC)\n SILVA Data")
names(volcanoTitleStrings) <- names(biomarkerPhyseqList)

volcanoPlotList <- pmap(.l = list(physeq = biomarkerPhyseqList,
                                  resultsDataTable = deseqResTableList,
                                  plotTitle = volcanoTitleStrings),
                        .f = PlotStaticVolcano,
                        sigThreshold = 0.05)

volcanoPlotList

#----- Save Static Volcano Plots -----#

volcanoPlotsPath <- paste0(figPath, "deseq_volcano_plots")

if (!file.exists(volcanoPlotsPath)) {
  dir.create(volcanoPlotsPath)
}

volcanoPlotNames <- paste0(volcanoPlotsPath, "/", names(biomarkerPhyseqList),
                           "-deseq-volcano.pdf")
walk2(.x = volcanoPlotList,
      .y = volcanoPlotNames,
      .f = ~ ggsave(filename = .y, plot = .x, width = 11, height = 8.5,
                    units = "in"))
```

```{r save-deseq-sig-results-ALL-COMPARISONS, eval=FALSE, include=FALSE}

#----- Filter Significant Results from deseqResTableList -----#

sigDESeqResTableList <- map(.x = deseqResTableList,
                            .f = ~ filter(.x, Significant == TRUE))

deseqResPath <- paste0(resultsPath, "/deseq2/all_comparisons/")

if (!file.exists(deseqResPath)) {
  dir.create(deseqResPath)
}

sigResFileNames <- paste0(deseqResPath, names(sigDESeqResTableList), 
                          "_DESeq2_SigRes.txt")

walk2(.x = sigDESeqResTableList,
      .y = sigResFileNames,
      .f = ~ write.table(x = .x, file = .y, quote = FALSE, sep = "\t",
                         row.names = FALSE))
```

```{r save-deseq-sig-results-OVERLAPS, eval=FALSE, include=FALSE}
preBiomarkerTable <- deseqResTableList$pre_timepoint %>% 
  filter(Significant == TRUE)
preBiomarkers <- preBiomarkerTable$OTU

postBiomarkerTable <- deseqResTableList$post_timepoint %>% 
  filter(Significant == TRUE)
postBiomarkers <- postBiomarkerTable$OTU

biomarkersUniqueToPost <- postBiomarkers[!postBiomarkers %in% preBiomarkers]
uniquePostBiomarkerTable <- postBiomarkerTable %>% 
  filter(OTU %in% biomarkersUniqueToPost)

biomarkersNotUniqueToPost <- postBiomarkers[postBiomarkers %in% preBiomarkers]
nonUniquePostBiomarkerTable <- postBiomarkerTable %>% 
  filter(OTU %in% biomarkersNotUniqueToPost)

# save pre and post biomarker tables, and biomarkers unique to post
deseqResPath <- paste0(resultsPath, "/", "deseq2")

if (!file.exists(deseqResPath)) {
  dir.create(deseqResPath)
}

deseqTableNames <- paste0(deseqResPath, "/overlaps/", c("post-biomarkers.txt",
                                                        "pre-biomarkers.txt",
                                                        "biomarkers-unique-to-post.txt"))

walk2(.x = list(postBiomarkerTable, preBiomarkerTable, uniquePostBiomarkerTable),
      .y = deseqTableNames,
      .f = ~ write.table(x = .x, file = .y, quote = FALSE, sep = "\t",
                         row.names = FALSE))
```


### Session Info

```{r session-info}
Sys.Date()
getwd()
sessionInfo()
```