---
title: "TINMAN WGS Analysis"
author: "Rachel Rodgers"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
runtime: shiny
---

## Overview {.tabset}

In this script I will be analyzing the bacterial microbiome using the data
generated from MetaPhlAn3.

Similar to the 16S analysis, we are interested in investigating changes in the 
gut  bacterial microbiome in patients who developed neutropenia after antibiotics
(10 patients), and comparing to matched controls (29 patients). Each patient has
a "pre" and "post" time point. This data set will be analyzed in several
different groupings:

  1. Neutropenic vs Controls in the full cohort (time points combined)
  2. Neutropenic vs Controls in the "pre" and "post" time points only
  3. "Pre" vs "Post" timepoints in the Control or Neutropenic group only
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=8,
                      fig.height=6,
                      warning = FALSE,
                      message = FALSE)

knitr::opts_knit$set(root.dir = "..")
```

```{r load-libraries, message=FALSE}
library("here")
library("shiny")
library("plotly")
library("phyloseq")
library("DESeq2")
library("microbiome")
library("data.table")
library("ggpubr")
library("ggrepel")
library("janitor")
library("vegan")
library("tidyverse")


source(here("shared_R_scripts/Helper_Functions.R"))
source(here("shared_R_scripts/BoxPlotHelpers.R"))
source(here("shared_R_scripts/BiomarkerHelpers.R"))

options(shiny.sanitize.errors = FALSE)

inputPath <- "../documents/original_documents/WGS/Deliverables/metaphlan3"
```

```{r physeqRawEstCounts, results='hide'}
# "Estimated Counts"
physeqRawEstCounts <- import_biom(here(file.path(inputPath,
                                                 "metaphlan3.bacteria.EstCount.biom")))
physeqRawEstCounts # 514 x 78
```

```{r modify-taxonomy-table, results='hide'}
# format column names and remove prefix strings in rank names
taxonomyTableOrig <- as.data.frame(as(tax_table(physeqRawEstCounts), "matrix"))

colnames(taxonomyTableOrig) <- c("Kingdom", "Phylum", "Class",
                                 "Order", "Family", "Genus",
                                 "Species")

# clean up strings
removePatterns <- paste0("^", c("k", "p", "c", "o", "f", "g", "s"), "_+")
finalPattern <- paste0(removePatterns, collapse = "|")

taxonomyTableModified <- taxonomyTableOrig %>% 
  mutate(across(everything(), ~ str_remove(string = .x,
                                           pattern = finalPattern)))
taxTableFinal <- as.matrix(taxonomyTableModified)

# add back to phyloseq object
physeqModifiedTax <- physeqRawEstCounts
physeqModifiedTax@tax_table <- NULL
physeqModifiedTax <- merge_phyloseq(physeqModifiedTax,
                                    tax_table(taxTableFinal))
```

```{r merge-sample-data, results='hide'}
sampleData <- readRDS(here("../data/RDataObjects/WGS/metadataFinal.RDS"))

# phyloseq sample names are slightly different due to hyphenation
phyloseqNames <- data.frame(phyloseqNames = sample_names(physeqRawEstCounts)) %>% 
  mutate(phyloseqNamesModified = str_remove_all(phyloseqNames, "-"))

sampleDataMerged <- base::merge(phyloseqNames, sampleData,
                                by.x = "phyloseqNamesModified",
                                by.y = "sample_name",
                                all = TRUE)

# add reads per sample from the phyloseq object
readsPerSample <- sample_sums(physeqRawEstCounts)

# make new column "timepoint_condition" that combines the pre_post and 
#   neutropenic column. this will be used for beta diversity plot
sampleDataMerged <- sampleDataMerged %>% 
  mutate("timepoint_condition" = paste(pre_post, neutropenic, sep = "_")) %>% 
  relocate(timepoint_condition, .after = "neutropenic")

sampleDataFinal <- sampleDataMerged %>% 
  mutate("readsPerSample" = readsPerSample[phyloseqNames])

row.names(sampleDataFinal) <- sampleDataFinal$phyloseqNames
```

```{r physeqBacteria, results='hide'}
physeqBacteria <- merge_phyloseq(physeqModifiedTax,
                                 sample_data(sampleDataFinal))
physeqBacteria # 514 taxa x 78 samples
```

```{r set-values-for-analyses}
# plotting colors
groupColors <- cbPaletteGrey[1:2]
names(groupColors) <- c("Control", "Neutropenic")

timepointColors <- c("PRE" = "blue", "POST" = "red")

tpCndColors <- cbPaletteGrey[5:8]
names(tpCndColors) <- unique(sampleDataFinal$timepoint_condition)

# Save columnNames to a reactiveValue as well so it can be accessed when 
#   running ADONIS calculations for beta diversity, and DESeq2 analysis
#   for biomarkers.
columnLUT <- reactiveValues("all_samples_by_cnd" = "neutropenic",
                            "all_samples_by_tp_cnd" = "timepoint_condition",
                            "pre_timepoint" = "neutropenic",
                            "post_timepoint" = "neutropenic",
                            "control_samples" = "pre_post",
                            "neutropenic_samples" = "pre_post")
```

### **QC**

<br>

<h4>**Read Count Distributions**</h4>

<br>

The samples will be compared in three different groupings. Therefore we will
first check for differences in the average reads/sample of each group, as any
significant differences could affect the outcome of the ecological measures.

```{r frequency-of-read-counts}
# make df of reads/sample
readsPerSampleDF <- sampleDataFinal %>% 
  select(phyloseqNames, neutropenic, pre_post, readsPerSample) %>% 
  arrange(readsPerSample)

# apply levels to samples so bar graph will show increasing reads/sample
sampleLevels <- pull(readsPerSampleDF, phyloseqNames)
readsPerSampleDF$phyloseqNames <- factor(readsPerSampleDF$phyloseqNames,
                                         levels = sampleLevels)
```

```{r summaryReadDist}
summary(readsPerSampleDF$readsPerSample)
```

```{r save-readsPerSampleDF-for-josie, include=FALSE, eval=FALSE}
write.table(readsPerSampleDF,
            file = "../analysis/results/WGS/WGSReadsPerSample.txt",
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)

```

Reads/sample range from 188,192 to 28,407,947 with an average of 12,557,903.

```{r readDistPlot}
readDistPlot <- ggplot(readsPerSampleDF,
                       aes(x = phyloseqNames, y = readsPerSample,
                           fill = neutropenic)) +
  scale_fill_manual(values = groupColors) +
  geom_bar(stat = "identity") +
  ggtitle("Read Count Distributions",
          subtitle = "Colored by Group") +
  theme_bw() +
  theme(axis.text.x = element_blank()) +
  guides(fill = guide_legend(title = "Group")) +
  facet_wrap(~ pre_post, scales = "free_x", nrow = 2, ncol = 1)

readDistPlot
```

Reads for subject 68 and 77 are lower than expected. Both the "pre" and "post"
samples for these two subjects will be removed from analysis.

```{r remove-low-read-samples}
sampleDataFinal <- filter(sampleDataFinal, !unique_id %in% c("068", "077"))

keepSamples <- sampleDataFinal$phyloseqNames
# b/f removal of 4 samples, physeqBacteria: 514 x 78
physeqBacteria <- prune_samples(samples = keepSamples, physeqBacteria) %>% 
  RemoveMissingTaxa() # 513 x 74
```

```{r readDistBox}
readDistBox <- PrettyBox(df = sampleDataFinal,
                         xVar = "neutropenic", yVar = "readsPerSample",
                         statMethod = "wilcox.test",
                         plotTitle = "Avg. Reads Across Groups (Time Points Combined)",
                         colorVals = groupColors,
                         label_x = "Group", label_y = "Avg. Reads/Sample")

readDistBox
```

<br>

There is no significant difference in average reads per sample between the 
control and neutropenic sample groups when combined across pre/post time points.

```{r readDistBoxByTime}
readDistBoxByTime <- PrettyBox(df = sampleDataFinal,
                               xVar = "neutropenic", yVar = "readsPerSample",
                               statMethod = "wilcox.test",
                               plotTitle = "Avg. Reads Across Groups Within Time Point",
                               colorVals = groupColors,
                               label_x = "Time Point", label_y = "Avg. Reads/Sample",
                               facet_formula = "~ pre_post",
                               facet_rows = 2, facet_cols = 1)

readDistBoxByTime
```

<br>

Likewise, there is no significant difference in the average number of 
reads per sample between the control and neutropenic samples at pre or post
time points.

<br>

```{r readDistBoxByGroup}
readDistBoxByGroup <- PrettyBox(df = sampleDataFinal,
                               xVar = "pre_post", yVar = "readsPerSample",
                               statMethod = "wilcox.test",
                               plotTitle = "Avg. Reads Across Time Points Within Groups",
                               colorVals = timepointColors,
                               label_x = "Group", label_y = "Avg. Reads/Sample",
                               facet_formula = "~ neutropenic",
                               facet_rows = 2, facet_cols = 1)

readDistBoxByGroup
```

<br>

When looking at the time points within the individual groups, there is no 
significant difference in average reads between the Pre and Post time points.
unevenly.

<br>
<br>

---

<h4>**Sample & Taxa Filtering**</h4>

<br>

Which bacterial taxa are present in the data set?

<br>

```{r phylaPrevalence}
phylaPrevalence <- TaxRankPrevalence(physeqBacteria, "Phylum")

phylaPrevalencePlot <- ggplot(phylaPrevalence$prevalence_df,
                              aes(x = TotalAbundance,
                                  y = Prevalence/nsamples(physeqBacteria),
                                  color = Family)) +
  geom_hline(yintercept = 1/nsamples(physeqBacteria), alpha = 0.5, 
             linetype = 2) +
  geom_point(size = 3, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") +
  ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~ Phylum) +
  theme_bw() +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  ggtitle("Phylum Prevalence in All Samples",
          subtitle = "Colored by Familiy")
phylaPrevalencePlot
```

```{r physeqSubsetList}
# subset physeqBacteria by:
#   all (physeqBacteria)
#   pre time point (pre_post == "PRE")
#   post time point (pre_post == "POST)
#   control group only (neutropenic == "Control")
#   neutropenic group only (neutropenic == "Neutropenic")

sampleDataBacteria <- as(sample_data(physeqBacteria), "data.frame")

subsetNames <- c("pre_timepoint",
                 "post_timepoint",
                 "control_samples",
                 "neutropenic_samples")

# What column to use to make each subset in subsetName:
columnNames <- c(rep("pre_post", 2), rep("neutropenic", 2))
names(columnNames) <- subsetNames

# What labels meet each condition:
conditionValues <- c("PRE",
                     "POST",
                     "Control",
                     "Neutropenic")
names(conditionValues) <- subsetNames

# Initialize list to hold the physeqSubsets
physeqSubsetList <- vector(mode = "list", length = length(subsetNames))

for (i in 1:length(subsetNames)) {

  currentSubset <- subsetNames[i]
  currentColumn <- columnNames[i]
  currentCondition <- conditionValues[i]
  
  keepSamples <- sampleDataBacteria %>% 
    filter(!!sym(currentColumn) == currentCondition) %>% 
    pull(phyloseqNames)
  
  psSubset <- prune_samples(keepSamples, physeqBacteria) %>% 
    RemoveMissingTaxa()
  
  physeqSubsetList[[i]] <- psSubset
  names(physeqSubsetList)[i] <- currentSubset
}

# Add all_samples as physeqBacteria
physeqSubsetList <- c("all_samples_by_cnd" = physeqBacteria, 
                      "all_samples_by_tp_cnd" = physeqBacteria, 
                      physeqSubsetList)
```

<br>
<br>

### **Community Composition**

<br>

Community composition plots can be used to examine the overall taxon representation
across samples.

```{r comm-comp}

#----- UI -----#

fluidRow(
  
  column(5,
         
         wellPanel(
           # select rank to display
           selectInput(inputId = "communityTaxRank", label = "Taxonomic Rank:",
                       choices = c("Kingdom", "Phylum", "Class", "Order",
                                   "Family", "Genus", "Species"),
                       selected = "Phylum"),
           
           # select abudnance filter
           sliderInput(inputId = "abdFilter", label = "Abundance Filter",
                       min = 0, max = 1, step = 0.05, value = 0),
           # update
           actionButton(inputId = "updateCommunity", 
                        label = "Update Community Plot",
                        icon = icon("fas fa-sync-alt"))
         ))
  
)

#----- React -----#

communityCompXLabs <- sampleDataFinal %>% 
  select(phyloseqNames, unique_id) %>%
  mutate(unique_id = str_remove(string = unique_id, "^0+")) %>% 
  deframe()

#test <- MakeAbundanceDF(physeqBacteria, "Phylum", 0)

# sampleLevels <- test %>% 
#       select(Sample, unique_id, pre_post) %>% 
#   distinct() %>% 
#   arrange(unique_id, pre_post)

GetCommunityRank <- reactive({
  
  rank <- input$communityTaxRank
  
  if (is.null(input$communityTaxRank)) {
    rank <- "Phylum"
  }

  return(rank)
  
})

CalculateAbundance <- eventReactive(input$updateCommunity, {
  
  withProgress(message = "Calculating Abundance...", value = 1, {
    
    abdDF <- MakeAbundanceDF(physeq = physeqBacteria,
                             taxRank = isolate(GetCommunityRank()),
                             abundanceFilter = isolate(input$abdFilter),
                             pruneMissing = FALSE)
    
    # apply levels to the Sample column of abdDF
    sampleLevels <- abdDF %>% 
      select(Sample, unique_id, pre_post) %>% 
      distinct() %>% 
      arrange(unique_id, pre_post) %>% 
      pull(Sample)
    
    abdDF$Sample <- factor(abdDF$Sample, levels = sampleLevels)
    
    if (dim(abdDF)[1] == 0) {
      stop("No taxa meet this filtering criteria. Try lowering the Abundance Filter option.",
           call. = FALSE)
    }
    
    if (isolate(GetCommunityRank()) == "Species") {
      abdDF <- mutate(abdDF, "GenusSpecies" = paste(Genus, Species, sep = " "))
    }
    
    return(abdDF)
    
  })
  
}, ignoreNULL = FALSE)

PlotCommunityCompositionModified <- function(abdDF, taxRank = "Phylum",
                                             facetFormula = NULL,
                                             facetCol = NULL, facetRow = NULL) {
  basePlot <- ggplot(abdDF,
                     aes_string(x = "Sample", y = "Abundance",
                                fill = taxRank)) +
    geom_bar(stat = "identity", width = 1, color = "grey14") +
    xlab("Subject") +
    ylab("Relative Abundance") +
    scale_x_discrete(labels = communityCompXLabs) +
    theme(axis.text.x = element_text(size = 12, angle = -90),
          plot.title = element_text(hjust = 0.5),
          plot.subtitle = element_text(hjust = 0.5))
  # are there facets??
  if (!(is.null(facetFormula))) {
    formula <- as.formula(facetFormula)
    facetPlot <- basePlot +
      facet_wrap(formula, scales = "free", nrow = facetRow, ncol = facetCol) +
      theme(panel.spacing.y = unit(1, "cm", data = NULL))
    return(facetPlot)
  } else {
    return(basePlot)
  }
  
}

renderPlotly({
  
  withProgress(message = "Loading plot...", value = 1, {
    
    compPlot <- PlotCommunityCompositionModified(abdDF = CalculateAbundance(),
                                                 taxRank = ifelse(
                                                   isolate(GetCommunityRank()) == "Species",
                                                   yes = "GenusSpecies",
                                                   no = isolate(GetCommunityRank())),
                                                 facetFormula = "pre_post ~ neutropenic",
                                                 facetCol = 2, facetRow = 2)
    ggplotly(compPlot, height = 800, width = 1000)
  
  })

})
```

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

### **Alpha Diversity**

<br>

Alpha diversity is concerned with describing the community within a sample. 
It is a standard tool used to calculate ecological indices which give the number 
of taxa present in a sample and the relationships between relative abundance and 
how evenly taxa are distributed. These indices provide useful summary 
information about the community structure of a given sample.

We are primarily interested in richness (the number of taxa per sample) and 
Shannon diversity (a description of the richness and evenness of a sample's 
community). An evenness of 0 indicates that a community is dominated by one or 
a few taxa, but an evenness of 1 means that species are evenly distributed. 
Diversity increases as richness and evenness increase.

```{r alphaDiv}
alphaIdx <- c("Richness" = "observed",
              "Peilou's Evenness" = "evenness_pielou",
              "Shannon Diversity" = "diversity_shannon")
alphaDiv <- microbiome::alpha(physeqBacteria,
                              index = alphaIdx)

sampleDataBacteria <- base::merge(alphaDiv, sampleDataBacteria, by = "row.names") %>% 
  column_to_rownames(var = "Row.names")
```

<br>
<br>

```{r PlotAlphaModified}
# Modified PrettyBox() function to add option for connecting samples by unique_id
#   and setting limits on the y-axes

PlotAlphaModified <- function(df, xVar, yVar, statMethod, plotTitle, yLims,
                              addLines = FALSE,
                              colorVals = NULL, label_x = NULL, label_y = NULL,
                              legendPos = "none", facet_formula = NULL,
                              facet_rows = NULL, facet_cols = NULL) {
  
  basePlot <- ggplot(data = df,
                     aes_string(x = xVar, y = yVar, fill = xVar)) +
    geom_boxplot(outlier.shape = NA) +
    scale_fill_manual(values = colorVals) +
    labs(x = label_x, y = label_y, title = plotTitle) +
    scale_y_continuous(limits = yLims) +
    theme_pubr() +
    theme(axis.title = element_text(size = 14),
          axis.text = element_text(size = 12),
          plot.title = element_text(size = 14, hjust = 0.5),
          strip.text = element_text(size = 12),
          legend.title = element_blank(), 
          legend.position = legendPos,
          legend.text = element_text(size = 12)) +
    stat_compare_means(method = statMethod, label.x.npc = 0)
  
  if (addLines) {
    basePlot <- basePlot +
      geom_point(size = 1.5) +
      geom_line(aes(group = unique_id))
  } else {
    basePlot <- basePlot +
      geom_jitter(width = 0.2, size = 1.5)
  }
  
  if (!is.null(facet_formula)) {
    basePlot <- basePlot +
      facet_wrap(as.formula(facet_formula), nrow = facet_rows, ncol = facet_cols,
                 scales = "free")
  }
  
  return(basePlot)
  
}
```

```{r alphaCombinedTP, fig.width=9, fig.height=4}
#----- Combined Time Points -----#

alphaCombinedTP <- pmap(.l = list(yVar = alphaIdx,
                                  plotTitle = names(alphaIdx)),
                        .f = PrettyBox,
                        df = sampleDataBacteria,
                        xVar = "neutropenic",
                        statMethod = "wilcox.test",
                        colorVals = groupColors)

alphaCombinedTPArr <- annotate_figure(ggarrange(plotlist = alphaCombinedTP,
                                                nrow = 1, ncol = 3),
                                      top = text_grob("Alpha Diversity (Time Points Combined)",
                                                      size = 14, face = "bold"))
alphaCombinedTPArr
```

<br>
<br>

---

<br>
<br>

```{r alphaSeparateTP, fig.width=9, fig.height=6}

#----- Separate Time Points -----#

alphaSeparateTP <- pmap(.l = list(yVar = alphaIdx,
                                  plotTitle = names(alphaIdx),
                                  yLims = list("Richness" = c(0, 180),
                                               "Peilou's Evenness" = c(0, 0.8),
                                               "Shannon Diversity" = c(0, 4))),
                        .f = PlotAlphaModified,
                        df = sampleDataBacteria,
                        xVar = "neutropenic",
                        statMethod = "wilcox.test",
                        colorVals = groupColors,
                        facet_formula = "~ pre_post",
                        facet_rows = 2, facet_col = 1)

alphaSeparateTPArr <- annotate_figure(ggarrange(plotlist = alphaSeparateTP,
                                                nrow = 1, ncol = 3),
                                      top = text_grob("Alpha Diversity at Pre and Post Time Points",
                                                      size = 14, face = "bold"))
alphaSeparateTPArr
```

<br>
<br>

---

<br>
<br>

```{r alphaPrePost, fig.width=9, fig.height=6}
alphaPrePost <- pmap(.l = list(yVar = alphaIdx,
                               plotTitle = names(alphaIdx),
                               yLims = list("Richness" = c(0, 180),
                                            "Peilou's Evenness" = c(0, 0.8),
                                            "Shannon Diversity" = c(0, 4))),
                     .f = PlotAlphaModified,
                     addLines = TRUE,
                     df = sampleDataBacteria,
                     xVar = "pre_post",
                     statMethod = "wilcox.test",
                     colorVals = timepointColors,
                     facet_formula = "~ neutropenic",
                     facet_rows = 2, facet_col = 1)


alphaPrePostArr <- annotate_figure(ggarrange(plotlist = alphaPrePost,
                                             nrow = 1, ncol = 3),
                                   top = text_grob("Alpha Diversity of Pre vs Post Time Points Within Groups",
                                                   size = 14, face = "bold"))
alphaPrePostArr
```

<br>
<br>

Richness is significantly higher in the "pre" time point for both the control and
neutropenic groups. This may be expected following antibiotics.

Evenness and Shanon Diversity are significantly higher in the "pre" time point 
for the neutropenic samples.

<br>
<br>

--

<br>
<br>

<h4> Change in Alpha Diversity Over Time: Control vs. Neutropenic </h4>

<br>
<br>

Here we are looking to see if the change in alpha diversity from the Pre to Post
timepoint is significantly different in the control samples versus the 
neutropenic samples. In this plot, each point represents the change in richness
for a given sample.

```{r calculate-change-in-alpha-diversity}
deltaMetrics <- list("Richness" = "observed",
                     "Shannon Diversity" = "diversity_shannon")

MakeDeltaDF <- function(metric) {
  
  newColName <- paste0("delta_", metric)
  
  deltaDF <- sampleDataBacteria %>%
    select(unique_id, !!sym(metric), pre_post, neutropenic) %>%
    pivot_wider(id_cols = c(unique_id, neutropenic),
                names_from = pre_post,
                values_from = !!sym(metric)) %>%
    mutate(!!newColName := abs(POST - PRE))

  
  return(deltaDF)
}

deltaAlphaDFList <- map(.x = deltaMetrics, .f = MakeDeltaDF)

deltaAlphaPlots <- pmap(.l = list(df = deltaAlphaDFList,
                                  yVar = list("Richness" = "delta_observed",
                                              "Shannon Diversity" = "delta_diversity_shannon"),
                                  label_y = paste("Absolute Change in",
                                                  names(deltaAlphaDFList)),
                                  plotTitle = paste("Change in",
                                                    names(deltaAlphaDFList),
                                                    "Over Time")),
                        .f = PrettyBox,
                        xVar = "neutropenic",
                        statMethod = "wilcox.test",
                        colorVals = groupColors)

deltaAlphaPlotsArr <- ggarrange(plotlist = deltaAlphaPlots,
                                nrow = 1, ncol = 2)
deltaAlphaPlotsArr
```

<br>
<br>

The average change in richness and Shannon diversity over time does not vary
significantly between control and neutropenic samples.

<br>
<br>

### **Beta Diversity**

<br>

While alpha diversity was concerned with describing the community within a 
sample, beta diversity compares the communities between samples. Beta diversity 
is quantified through "association" coefficients - similarity or distance 
measures. It describes how similar/dissimilar samples' communities are to one 
another. We view these distances with ordination plots such as PCoA.

Because this data set lacks a phylogenetic tree, I will be using the Bray-Curtis
dissimilarity metric to quantify the compositional dissimilarity between
communities. This metric is based on the counts of different species in each 
group being compared.


<br>
<br>

```{r beta-diversity-ui}
fluidRow(
  column(5,
         wellPanel(
           # select group to plot
           selectInput(inputId = "betaGroup",
                       label = "Select Group",
                       choices = c("All (by condition)" = "all_samples_by_cnd",
                                   "All (by timepoint & condition)" = "all_samples_by_tp_cnd",
                                   "PRE Time Point" = "pre_timepoint",
                                   "POST Time Point" = "post_timepoint",
                                   "Control" = "control_samples",
                                   "Neutropenic" = "neutropenic_samples"),
                       selected = "All (by condition)"),
           # show ellipses?
           checkboxInput(inputId = "betaEllipsesBool",
                         label = "Show Confidence Ellipses",
                         value = FALSE),
           # show lines connecting pre/post samples
           checkboxInput(inputId = "betaShowLines",
                         label = "Show Lines Connecting Subject Samples",
                         value = FALSE),
           # update
           actionButton(inputId = "betaUpdate",
                        label = "Update Beta Diversity Plots",
                        icon = icon("fas fa-sync-alt"))
         ))
)
```

```{r beta-diversity-reactive}

#----- Handle Start Up NULLS -----#

SelectedPhyseq <- reactive({
  selectedPhyseq <- input$betaGroup
  if (is.null(selectedPhyseq)) {
    selectedPhyseq <- "all_samples_by_cnd"
  }
  return(selectedPhyseq)
})


#----- Calculate Ordination -----#

CalculateOrdination <- eventReactive(input$betaUpdate, {
  withProgress(message = "Calculating Ordination...", value = 1, {
    set.seed(11643764)
    ordinationObj <- ordinate(physeqSubsetList[[SelectedPhyseq()]],
                              method = "PCoA",
                              distance = "bray")
  })
}, ignoreNULL = FALSE)

#----- Run ADONIS for Significance -----#

AdonisPVal <- eventReactive(input$betaUpdate, {
  set.seed(11643764)
  adonisRes <- RunAdonis(physeqSubsetList[[SelectedPhyseq()]],
                         category = columnLUT[[SelectedPhyseq()]],
                         distance = "bray")
  pVal <- adonisRes$aov.tab$`Pr(>F)`[1]
}, ignoreNULL = FALSE)


#----- Render the Ordination Plot -----#

observe({print(SelectedPhyseq())})

MakeOrdPlot <- eventReactive(input$betaUpdate, {
  
  colorOptions <- list("all_samples_by_cnd" = groupColors,
                       "all_samples_by_tp_cnd" = tpCndColors,
                       "pre_timepoint" = groupColors,
                       "post_timepoint" = groupColors,
                       "control_samples" = timepointColors,
                       "neutropenic_samples" = timepointColors)
  
  titleOptions <- c("all_samples_by_cnd" = "All Samples",
                    "all_samples_by_tp_cnd" = "All Samples",
                    "pre_timepoint" = "Pre Time Point Samples",
                    "post_timepoint" = "Post Time Point Samples",
                    "control_samples" = "Control Samples",
                    "neutropenic_samples" = "Neutropenic Samples")
  
  ordPlot <- plot_ordination(physeq = physeqSubsetList[[SelectedPhyseq()]],
                             ordination = CalculateOrdination(),
                             color = columnLUT[[SelectedPhyseq()]]) +
    scale_color_manual(values = colorOptions[[SelectedPhyseq()]]) +
    theme_bw() +
    geom_point(size = 3.5) +
    ggtitle(paste("Bray Distance of", 
                  titleOptions[[SelectedPhyseq()]]),
            subtitle = paste("ADONIS p-val:", AdonisPVal())) +
    theme(plot.title = element_text(hjust = 0.5, size = 16),
          plot.subtitle = element_text(hjust = 0.5, size = 14),
          legend.title = element_text(size = 16),
          legend.text = element_text(size = 14))
  
  # Show ellipses?
  if (input$betaEllipsesBool) {
    ordPlot <- ordPlot + stat_ellipse(type = "norm")
  }
  
  # Show lines?
  if (input$betaShowLines) {
    ordPlot <- ordPlot +
      geom_line(aes_string(group = "unique_id"))
  }
  
  return(ordPlot)
  
}, ignoreNULL = FALSE)


renderPlot({
  
  MakeOrdPlot()
  
}, height = 600, width = 900)
```

<br>
<br>

According to ADONIS, post time point neutropenic and control sample communities 
are significantly dissimilar. Additionally, pre vs post control samples 
communities are significantly dissimilar.


<br>
<br>

### **Biomarker Analysis with DESeq2**

<br>

We will use DESeq2 to investigate potential biomarkers between groups. This 
information may be referenced against the community composition plots to
identify broad taxonomic groups that may identify a subset of samples.

<br>
<br>

```{r GenerateDESeqResultsModified}
GenerateDESeqResultsModified <- function(physeq, variable, numerator, denominator) {
  
  # Returns DESeq Results as Formal Class "DESeqResults"
  # Create formula from string variable
  formula <- as.formula(paste("~", variable, sep = " "))
  # Convert to deseq data set object
  ds <- phyloseq_to_deseq2(physeq, design = formula)
  # work-around
  counts <- counts(ds)
  geoMeans <- apply(counts, 1,
                      function(row) {if (all(row == 0)) 0 else exp(mean(log(row[row != 0])))})
  dds <- estimateSizeFactors(ds, geoMeans = geoMeans)
  # Run analysis
  ddsAnalysis <- DESeq(dds, test = "Wald", fitType = "local", betaPrior = FALSE)
  # Extract and format results
  ddsResults <- results(ddsAnalysis,
                        contrast = c(variable, numerator, denominator)) 
}
```

```{r set-reactiveValues-for-biomarker-analyses}

numeratorLUT <- reactiveValues("all_samples_by_cnd" = "Control",
                               "pre_timepoint" = "Control",
                               "post_timepoint" = "Control",
                               "control_samples" = "PRE",
                               "neutropenic_samples" = "PRE")

denominatorLUT <- reactiveValues("all_samples_by_cnd" = "Neutropenic",
                                 "pre_timepoint" = "Neutropenic",
                                 "post_timepoint" = "Neutropenic",
                                 "control_samples" = "POST",
                                 "neutropenic_samples" = "POST")
```

```{r biomarker-ui}
fluidRow(
  column(5,
         wellPanel(
           # select group to plot
           selectInput(inputId = "biomarkerComparison",
                       label = "Select Group",
                       choices = c("All" = "all_samples_by_cnd",
                                   "PRE Time Point" = "pre_timepoint",
                                   "POST Time Point" = "post_timepoint",
                                   "Control" = "control_samples",
                                   "Neutropenic" = "neutropenic_samples"),
                       selected = "All"),
           # update
           actionButton(inputId = "updateBiomarkers",
                        label = "Run DESeq2",
                        icon = icon("fas fa-sync-alt"))
         ))
)
```

```{r biomarker-reactive}
# Set initial value of biomarkerComparison
BiomarkerComparison <- reactive({
  biomarkerComparison <- input$biomarkerComparison
  
  if (is.null(biomarkerComparison)) {
    biomarkerComparison <- "all_samples_by_cnd"
  }
  
  return(biomarkerComparison)
})

RunDESeq <- eventReactive(input$updateBiomarkers, {
  
  withProgress(message = "Conducting DESeq2 Analysis...", value = 1, {
    
    colName <- columnLUT[[BiomarkerComparison()]]
    numeratorVar <- numeratorLUT[[BiomarkerComparison()]]
    denominatorVar <- denominatorLUT[[BiomarkerComparison()]]
    
    deseqRes <- GenerateDESeqResultsModified(physeq = physeqSubsetList[[BiomarkerComparison()]],
                                             variable = colName,
                                             numerator = numeratorVar,
                                             denominator = denominatorVar)
  })
  
}, ignoreNULL = FALSE)

DESeqRes <- eventReactive(input$updateBiomarkers, {
  
  deseqResTable <- GenerateDESeqResultsTable(physeq = physeqSubsetList[[BiomarkerComparison()]],
                                             sigThreshold = 0.05,
                                             ddsResults = RunDESeq())
}, ignoreNULL = FALSE)
```

<br>

```{r biomarker-plot}
renderPlotly({
  volcano <- PlotStaticVolcano(physeq = physeqSubsetList[[BiomarkerComparison()]],
                               resultsDataTable = DESeqRes(),
                               sigThreshold = 0.05,
                               plotTitle = paste("Differentially Abundant Taxa\n",
                                                 denominatorLUT[[isolate(BiomarkerComparison())]],
                                                 "(-L2FC) vs",
                                                 numeratorLUT[[isolate(BiomarkerComparison())]],
                                                 "(+ L2FC)\n WGS Data"))
  
  ggplotly(volcano, tooltip = c("Phylum", "Family", "Genus", "Species",
                                "log2FoldChange", "baseMean"),
           height = 550, width = 900)
})
```

<br>
<br>
<br>
<br>

### Session Info

```{r session-info}
Sys.Date()
getwd()
sessionInfo()
```
